Source/WebCore/rendering/RenderLayerCompositor.cpp

```diff
@@ -333,47 +338,109 @@ void RenderLayerCompositor::BackingSharingState::startBackingSharingSequence(Ren
         candidate.providerLayer->backing()->setBackingSharingLayers(WTFMove(candidate.sharingLayers));
     }
-    m_backingProviderCandidates.clear();
     m_backingSharingStackingContext = nullptr;
 
     issuePendingRepaints();
 }
 
-auto RenderLayerCompositor::BackingSharingState::backingProviderCandidateForLayer(const RenderLayer& layer) -> Provider*
+auto RenderLayerCompositor::BackingSharingState::backingProviderCandidateForLayer(const RenderLayer& layer, const RenderLayerCompositor& compositor, LayerOverlapMap& overlapMap, OverlapExtent& overlap) -> Provider*
 {
     if (layer.hasReflection())
         return nullptr;
 
-    for (size_t i = 0; i < m_backingProviderCandidates.size(); ++i) {
-        auto& candidate = m_backingProviderCandidates[i];
+    if (!m_allowOverlappingProviders) {
+        for (size_t i = 0; i < m_backingProviderCandidates.size(); ++i) {
+            auto& candidate = m_backingProviderCandidates[i];
+            auto& providerLayer = *candidate.providerLayer;
+            if (layer.ancestorLayerIsInContainingBlockChain(providerLayer))
+                return &candidate;
+        }
+
+        return nullptr;
+    }
+
+    if (!m_backingProviderCandidates.size())
+        return nullptr;
+
+    LOG_WITH_STREAM(Compositing, stream << "Looking for backing provider candidate for " << &layer);
+
+    // First, find the frontmost provider that is an ancestor in the containing block chain.
+    int32_t candidateIndex = m_backingProviderCandidates.size() - 1;
+    bool foundPreviousCandidate = false;
+    for (; candidateIndex >= 0; --candidateIndex) {
+        auto& candidate = m_backingProviderCandidates[candidateIndex];
         auto& providerLayer = *candidate.providerLayer;
 
-        if (layer.ancestorLayerIsInContainingBlockChain(providerLayer))
-            return &candidate;
+        if (&layer != &providerLayer && layer.ancestorLayerIsInContainingBlockChain(providerLayer)) {
+            LOG_WITH_STREAM(Compositing, stream << "Found " << &providerLayer);
+            if (!foundPreviousCandidate)
+                return &candidate;
+            break;
+        }
+
+        LOG_WITH_STREAM(Compositing, stream << "Rejected " << &providerLayer);
+        foundPreviousCandidate = true;
     }
 
-    return nullptr;
+    if (candidateIndex < 0)
+        return nullptr;
+
+    auto& candidate = m_backingProviderCandidates[candidateIndex];
+    auto& providerLayer = *candidate.providerLayer;
+
+    LayoutRect overlapBounds;
+    LOG_WITH_STREAM(Compositing, stream << "Provider: composited scroll(" << providerLayer.canUseCompositedScrolling() << ") scrollableArea(" << providerLayer.scrollableArea() << ") horizontaloverflow(" << (providerLayer.scrollableArea() ? providerLayer.scrollableArea()->hasScrollableHorizontalOverflow() : false) << ")");
+    if  (providerLayer.canUseCompositedScrolling() && providerLayer.scrollableArea() && !providerLayer.scrollableArea()->hasScrollableHorizontalOverflow()) {
```

Comment:
It's weird to do this optimization only for vertical overflow. Why not also handle horizontal-only overflow? This could affect many carousels.